frame/bridge/relay-authorities/src/lib.rs

```diff
@@ -0,0 +1,615 @@
+		//
+		// This operation is forced to set the authorities,
+		// without the authorities change signature requirement
+		#[weight = 10_000_000]
+		pub fn reset_authorities(origin, authorities: Vec<RelayAuthorityT<T, I>>) {
+			T::ResetOrigin::ensure_origin(origin)?;
+
+			ensure!(!Self::on_authorities_change(), <Error<T, I>>::OnAuthoritiesChangeDis);
+
+			<Authorities<T, I>>::mutate(|old_authorities| {
+				for authority in old_authorities.iter() {
+					<RingCurrency<T, I>>::remove_lock(
+						T::LockId::get(),
+						&authority.account_id
+					);
+				}
+
+				*old_authorities = authorities;
+			});
+		}
+
+		// No-op if already submit
+		#[weight = 10_000_000]
+		pub fn submit_mmr_root_signature(
+			origin,
+			mmr_root: MMRRoot<T>,
+			signature: RelaySignature<T, I>
+		) {
+			let authority = ensure_signed(origin)?;
+
+			ensure!(!Self::on_authorities_change(), <Error<T, I>>::OnAuthoritiesChangeDis);
+
+			let (deadline, mut signatures) =
+				<SignedMMRRoots<T, I>>::get(&mmr_root).ok_or(<Error<T, I>>::MMRRootNE)?;
+
+			if signatures.iter().position(|(authority_, _)| authority_ == &authority).is_some() {
+				return Ok(());
+			}
+
+			let authorities = <Authorities<T, I>>::get();
+			let signer = find_signer::<T, I>(
+				&authorities,
+				&authority
+			).ok_or(<Error<T, I>>::AuthorityNE)?;
+
+			ensure!(
+				T::Sign::verify_signature(&signature, mmr_root, signer),
+				 <Error<T, I>>::SignatureInv
+			);
+
+			signatures.push((authority, signature));
+
+			if Perbill::from_rational_approximation(signatures.len() as u32 + 1, authorities.len() as _)
+				>= T::SignThreshold::get()
+			{
+				<MMRRootsToSign<T, I>>::mutate(|mmr_roots_to_sign|
+					if let Some(position) = mmr_roots_to_sign
+						.iter()
+						.position(|mmr_root_| mmr_root_ == &mmr_root)
+					{
+						mmr_roots_to_sign.remove(position);
+					}
+				);
+				<SignedMMRRoots<T, I>>::remove(&mmr_root);
+				<ClosedMMRRootSubmits<T, I>>::remove(&deadline);
+
+				// TODO: clean the mmr root which was contains in this mmr root?
+
+				Self::deposit_event(RawEvent::SignedMMRRoot(mmr_root, signatures));
+			} else {
+				<SignedMMRRoots<T, I>>::insert(&mmr_root, (deadline, signatures));
+			}
+		}
+
+		// No-op if already submit
+		#[weight = 10_000_000]
+		pub fn submit_authorities_signature(origin, signature: RelaySignature<T, I>) {
+			let old_authority = ensure_signed(origin)?;
+
+			ensure!(Self::on_authorities_change(), <Error<T, I>>::OnAuthoritiesChangeDis);
+
+			let (hashed_authorities_set, mut signatures) = <AuthoritiesToSign<T, I>>::get();
+
+			if signatures
+				.iter()
+				.position(|(old_authority_, _)| old_authority_ == &old_authority)
+				.is_some()
+			{
+				return Ok(());
+			}
+
+			let old_authorities = <OldAuthorities<T, I>>::get();
+			let signer = find_signer::<T, I>(
+				&old_authorities,
+				&old_authority
+			).ok_or(<Error<T, I>>::AuthorityNE)?;
+
+			ensure!(
+				T::Sign::verify_signature(&signature, hashed_authorities_set, signer),
```

Comment:
The cases of historical hashed_authorities_set duplicate and replay attack.

Introducing nonce?